### ISSUE TITLE
Unprotected pull from Docker Registry V2

### ISSUE THREAD

**Issue by gauravbhandare** [docker/distribution#1028]
I have V1 registry with nginx reverse proxy. My configuration allows docker pull without login and for any push login is needed. This all is working great till I started migrating to V2. With migration to V2 version of registry, i found below issues
- GET requests also need to be protected for docker login to work
- If I try to do a pull with every HTTP method protected for V2 registry and not logging in, the pull falls back to v1 as it gets 401 from /v2/ call instead of prompting for login. Image gets pulled from V2 registry perfectly if i do docker login manually and then do a pull.

Docker Registry Version - 2.1.1
Docker Version - 1.7.1
NGINX Version -  1.8.0

Anyone else facing similar issues with migration?

**Comment by RichardScothern** [comment]
The V2 registry works somewhat differently from v1.  The recommended setup for putting a V2 registry behind nginx is detailed here:
https://github.com/docker/distribution/blob/master/docs/nginx.md

**Comment by gauravbhandare** [comment]
I have the registry running. As I am using same nginx instance for V1 and V2, the pull request falls back to v1 when v2 call returns 401 instead if prompting for login credentials. The image does exist in v2 registry. As GET/HEAD calls cannot be unprotected for V2 it makes every single call fall back to V1 unless user manually runs docker login before running the pull command. 

Please let me know if this explains the scenario better.

**Comment by mv012004** [comment]
I ran into the same thing, any one has any workaround or ideas for the question asked by @gauravbhandare

**Comment by mrwacky42** [comment]
Until the client is made smart enough to tell the difference between 401 and 404, my solution is for Nginx to add this header:

            add_header 'Docker-Distribution-Api-Version' 'registry/2.0' always;

After auth, if you inspect the HTTP traffic, you'll see this header twice, because the registry adds it. It was too much effort to get Nginx to ensure it's only there once.

**Comment by ballerabdude** [comment]
@mrwacky42 Dude you saved my life. About 4 days of very little sleep my private docker registry is up and running. If anyone wants an AWS cloudformation stack that is minimal and will cost $7/month to run your own docker registry let me know.

**Comment by RichardScothern** [comment]
I am closing this issue.  If the header requirements are not clearly enough stated in the linked document, we always welcome PRs to improve clarity.

@ballerabdude : please refrain from advertising your business in our github issue tracker.

**Comment by mv012004** [comment]
@RichardScothern i don't understand why the issue would be closed ? I feel this issue is still open, as pointed out by @gauravbhandare, i am running into the same thing. Having said that, this is not a documentation issue, but something that isn't working in the docker-registry v2, the way it uses GET/HEAD requests for authentication while doing a PUSH image. Is there a solution/workaround for this.!

**Comment by mrwacky42** [comment]
@RichardScothern - I think he meant to say he's got a CF stack to spin up a registry that will cost $END_USER $7/month to run on AWS, not that he's providing that as a service.

As far as a solution, I think it's teaching the docker client that `401` is not a good reason to fall back to v1..

**Comment by RichardScothern** [comment]
@mrwacky42 : it's off topic for the issue at hand.   

@mv012004 did you follow the instructions in the linked document?

**Comment by gauravbhandare** [comment]
@RichardScothern i tried the NGINX configuration as mentioned by the link you provided. The same did not work as well and same issue persists. Its more because of the way Docker client handles the responses from V2 registry and falls back to V1. My current setup is NGINX fronted for Registry V1 and allows unprotected pulls but prompts for login for push requests. When i have Registry V2 also behind the same NGINX, since V2 has no way to unprotect only pull requests, if i protect pull for V2 then instead of prompting for login it falls back to V1 registry. Here is the complete nginx conf that I am using and works perfectly fine with Docker Registry V1. For V2, i face the issue mentioned in the issue description.

```
#user  nginx;
worker_processes  2;

error_log  /nginx01/error.log warn;

events {
    worker_connections  1024;
}


http {
    include       mime.types;
    default_type  application/octet-stream;
    log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';

    access_log  /nginx01/access.log main;
    sendfile        on;
    keepalive_timeout  65;

    
    upstream docker-registry {
        server <Docker Registry V1 Primary host:port>;
        server <Docker Registry V1 Backup host:port> backup;
    
    }
    upstream docker-registry-v2 {
        server <Docker Registry V2 Primary host:port>;
        server <Docker Registry V2 Backup host:port> backup;
    }


    ldap_server ldapserver {
        url "<LDAP URL>";
        binddn "<BINDDN>";
        binddn_passwd "<BINDDN_PASSWORD>";
        group_attribute member;
        group_attribute_is_dn on;
        satisfy any;
        require valid_user;
    }
    server {
        listen       <PORT>;
        proxy_set_header  Host              $http_host;   # required for docker client's sake
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Real-IP $remote_addr;
    
        proxy_set_header X-Forwarded-Proto https;
        proxy_set_header X-Original-URI $request_uri;

        client_max_body_size 0; # disable any limits to avoid HTTP 413 for large image uploads

        # required to avoid HTTP 411: see Issue #1486 (https://github.com/dotcloud/docker/issues/1486)
        chunked_transfer_encoding on;


    location / {
        location /v1 {
            limit_except GET {
                auth_ldap "Restricted Space";
                auth_ldap_servers ldapserver;
            }
            
            proxy_pass http://docker-registry;
        }
                location /v2 {
                    add_header Docker-Distribution-Api-Version 'registry/2.0' always;
                        limit_except GET {
                                auth_ldap "Restricted Space";
                                auth_ldap_servers ldapserver;
                        }
                        proxy_pass http://docker-registry-v2;
                }
        }

        error_page  404              /errors/error.html;

        # redirect server error pages to the static page /50x.html
        #
        #error_page   500 502 503 504  /50x.html;
        #location = /50x.html {
        #    root   html;
        #}
    }
}
```

**Comment by gauravbhandare** [comment]
I did some more debugging and i guess the issue is because of the way Docker client initiates the calls for V1 and V2 registry versions
For V1, if user does a docker push, the first call is made with PUT/POST for /v1/repositories/xxx and as the same is protected the client gets 401 back and is prompted for login whereas for V2 the first call is GET /v2/ which returns 200 as my config has unprotected pull and the docker client considers that it can start pushing manifests and the PUT fails with 401. If first call is PUT/POST for docker push i guess everything will work as expected with Docker Registry V2 as well.

My overall intention is to have unprotected pulls so that anyone can run a container using the image from repository but any push needs to be protected so that only authorized users are pushing the upgrades.

Hope this gives better understanding.

**Comment by mv012004** [comment]
@gauravbhandare - I scanned through the logs and observed the exact same behavior. I need to secure the push with authentication/authorization but want to allow everyone to pull images irrespective of doing docker login prior to pull, which makes sense and fits the same case as pulling images from Docker hub.

**Comment by gauravbhandare** [comment]
Little more debugging seems to point that the /auth on docker daemon is having some issue when the registry is configured with unprotected GET and returns 200 for /v2/ during docker login command execution. 

Logs from Docker daemon
```
DEBU[0411] Calling POST /auth
INFO[0411] POST /v1.20/auth
DEBU[0411] hostDir: /etc/docker/certs.d/<REGISTRY_HOST:PORT>
DEBU[0411] pinging registry endpoint https://<REGISTRY_HOST:PORT>/v0/
DEBU[0411] attempting v2 ping for registry endpoint https://<REGISTRY_HOST:PORT>/v2/
DEBU[0412] attempting v2 login to registry endpoint https://<REGISTRY_HOST:PORT>/v2/
ERRO[0412] Handler for POST /auth returned error: no successful auth challenge for https://<REGISTRY_HOST:PORT>/v2/ - errors: []
ERRO[0412] HTTP Error                                    err=no successful auth challenge for https://<REGISTRY_HOST:PORT>/v2/ - errors: [] statusCode=500
````

Logs from the access log for NGINX fronting Registry V2
```
<IP> - - [17/Oct/2015:18:27:02 -0400] "GET /v2/ HTTP/1.1" 200 2 "-" "docker/1.8.2 go/go1.4.2 git- os/linux arch/amd64" "-"
````

So even if the registry is returning 200, the handling in auth.go seems to give 500 internal error even before the "trying "basic" auth challenge with params" log entry.

When the registry is protected for GET as well, the client logs show clean login.
```
DEBU[0368] Calling POST /auth
INFO[0368] POST /v1.20/auth
DEBU[0368] hostDir: /etc/docker/certs.d/<REGISTRY_HOST:PORT>
DEBU[0368] pinging registry endpoint https://<REGISTRY_HOST:PORT>/v0/
DEBU[0368] attempting v2 ping for registry endpoint https://<REGISTRY_HOST:PORT>/v2/
DEBU[0368] attempting v2 login to registry endpoint https://<REGISTRY_HOST:PORT>/v2/
DEBU[0368] trying "basic" auth challenge with params map[realm:Docker Registry - Restricted Space v2 scope:]

```

Trying to pinpoint the issue but hoping for some expect to take a look as well.

**Comment by gauravbhandare** [comment]
This needs adding new POST end point in Registry like V1 had /v1/users and using the same for login on client side instead of /v2/

**Comment by mv012004** [comment]
This is a serious issue, with the docker client and probably need to be opened under docker-client however this needs to be taken look at, i am not sure why there is no activity on this issue. Can some of the devs look at this.

The issue seems to be with the docker login call doing a GET which is unprotected and since the GETs are allowed it assumed login successful even though the login has never happened, this needs to be fixed :|

**Comment by hjacobs** [comment]
@mv012004 it's a serious issue for us too: https://github.com/zalando-stups/pierone/issues/57

**Comment by danielmotaleite** [comment]
So, if i understand correctly, docker right now can not do unauthenticated pulls from registry and have authentication enable only in the push, correct? it must be "auth enable" for everything or "auth disabled" for everything.

As a workaround, one can create 2 sites configs, one for ```GET``` and ```HEAD``` requests (all other requests are denied) and unauthenticated. The other site config, using a different dns, allows every request and is always request authentication. Both of course share the same registry backend. This allows one public registry where only authenticated users can push, via a different address. Example:

dockerhub-ro.example.net - GET and HEAD without auth
dockerhub-rw.example.net - everything with auth

Not perfect but may help some people, specially since this thick had almost 1 year without any activity.

As a long term fix, docker should be fixed to be able to authenticate both PULL and PUSH requests independently of the other

**Comment by hjacobs** [comment]
@danielmotaleite yes, we are using exactly this workaround with two DNS domains pointing to the same registry and the server changing its behavior depending on `Host` header:

* docker pull registry.opensource.zalan.do/stups/openjdk:latest
* docker push registry-write.opensource.zalan.do/stups/openjdk:someversion

**Comment by SpeedBlack** [comment]
Hello @danielmotaleite,

Can you tell me how to use this workaround ?

I don't use nginx. I create two registry (2 different addresses / same registry backend) :
- 1 registry read-only mode, no authentication
- 1 registry RW, basic authentication

Do you think this method can cause problems ? Conflict between the 2 registry.

Thansk,
Best regards

**Comment by danielmotaleite** [comment]
@SpeedBlack what i have done is having ONE registry, being publish via 2 reverse-proxy, using 2 DNS

you say you have  "2 registry", but then say "same registry backend" , so i assume that the first reference, the "2 registry" is really a "reverse proxy"... i use nginx for reverse proxy but almost all reverse proxy servers that support authentication would work (haproxy, apache, nginx, etc)

If the registry backend is really the same, that is fine, that is the root solution from this workaround, it's just being presented in 2 different ways

**Comment by SpeedBlack** [comment]
Hello @danielmotaleite,

Ok, thanks ! After my tests it seems to work properly.
About the configuration, I do not know if I should use the same value for "REGISTRY_HTTP_SECRET" option.

In the documentation, it's written about "REGISTRY_HTTP_SECRET" : 

> A random piece of data used to sign state that may be stored with the client to protect against tampering. For production environments you should generate a random piece of data using a cryptographically secure random generator. If you omit the secret, the registry will automatically generate a secret when it starts. If you are building a cluster of registries behind a load balancer, you MUST ensure the secret is the same for all registries.

have you any idea ?
Thanks !

**Comment by danielmotaleite** [comment]
You need to set the `REGISTRY_HTTP_SECRET` when you have several backend servers working in a cluster. Having only one backend with multiple load-balancer sites will not need this.

**Comment by SpeedBlack** [comment]
Yes, I have installed 2 registries and both using the same storage:

Docker registry read only :
```
docker run -d 
-p 5001:5000 
–restart=always 
–name registry-ro 
-e REGISTRY_HTTP_HOST=registry-ro.example.com 
-v /registry/TEST_RO_RW/:/var/lib/registry:ro
.....
.....
registry:2
```

Docker registry Read-write :
```
docker run -d 
-p 5002:5000 
–restart=always 
–name registry-rw 
-e REGISTRY_HTTP_HOST=registry-rw.example.com 
-v /registry/TEST_RO_RW/:/var/lib/registry
.....
..... 
registry:2
```

You write:

> My problem with your setup is what happens when someone is uploading something and at same time, someone is reading the same image, but i have not enough knowledge about how registry works to guess. Try asking the developers and/or test your self

I tried multiple scenarios and it seems ok but can you explain to me exactly ?
I don't understand very well and I would like to trie.

Thanks a lot !

**Comment by alparslanozturk** [comment]
this way is not good idea because pull / push use url/imagename:version so we need layer7 routing .so nginx maybe do thant. or the private registry software has a option. but I don't know
